metric: add cidr based routing

andrewbaptist opened the pull request:
Add CIDR based lookup cache and modify the sql metrics for bytes in and out to become agg metrics rather than regular counter metrics.

Epic: CRDB-41142

Release note:
Modifies metrics sql.bytesin and sql.bytesout to be agg metrics if child metrics are enabled.
Adds a new configuration parameters
server.cidr_mapping_url which maps IPv4 CIDR blocks to arbitrary tag
names.

Commit:
util: add utility to process CIDR records

This commit adds a new utility which can store and efficiently process a
large number of CIDR records by mapping them to a unique name.

Epic: CRDB-41142

Release note: None

Commit:
sql: restructure sql byte metrics to be add metrics

This commit changes the SQL byte metrics to be broken down by cidr block
of the source.

Epic: CRDB-41142

Release note (ops change): Modifies metrics sql.bytesin and sql.bytesout
to be agg metrics if child metrics are enabled.

Commit:
sql: propagate the cidr.Lookup into ExecutorConfig

This commit constructs the cidr.Lookup into the sql server
ExecutorConfig and evalContext. Additionally this enables the cidr
mapping and adds a new configuration parameter for it.

Epic: CRDB-41142

Release note (ops change): Adds a new configuration parameters
server.cidr_mapping_url which maps IPv4 CIDR blocks to arbitrary tag
names.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/128707)
<!-- Reviewable:end -->

andrewbaptist commented:
This extracts from the other PR to only include the SQL metrics. I'll make one more pass on it before marking ready for review.

andrewbaptist commented on pkg/util/cidr/cidr.go:
```diff
@@ -0,0 +1,328 @@
+// transformJSON is an utility function which converts the contents of the
+// byte[] into CIDR records. It fails fast if the json is malformed. The
+// document must be of the format [{Name: "name", Ipnet: "ipnet"}, ...]
+func (c *Lookup) transformJSON(
```
Good point! I had originally made this more complicated with additional transforms of the JSON, but none of that is necessary anymore with the current version. I added tests for this and simplified to only call Unmarshall. My understanding of the docs is that Unmarshall should never panic, however I'll do a bit more testing.

andrewbaptist commented on pkg/util/cidr/cidr.go:
```diff
@@ -0,0 +1,328 @@
+// through all possible lengths from shortest to longest and checking if the
+// prefix matches.
+func (c *Lookup) LookupIP(ip net.IP) string {
+	if c == nil {
```
This isn't common, and I was a little torn on this. I think with the last changes I made I can remove this. I removed these checks for now and will fix the tests where this is not set up.

andrewbaptist commented on pkg/util/cidr/cidr.go:
```diff
@@ -0,0 +1,328 @@
+	contents := []byte("[]")
+	var err error
+	// Check if the URL is a local file
+	if strings.HasPrefix(rawURL, "file://") {
```
This is a valid point and it would definitely be a concern for longer lived connections like gRPC. For the primary connections it covers, I'm not as concerned. The SQL connections typically have a lifetime measured in minutes and the backup connection is only for the life of the backup operations. Changefeeds may have long-lived connections to Kafka, so it could matter there, but I'm not sure....

Currently the Start function blocks on initial refresh. I'm not sure if that is ideal, and it would be possible to move this onto the cidr-refresh goroutine fairly easily as well. This might be what happens in practice anyway since the inital call will not have the mapping URL set.

JeffSwenson replied:
>The SQL connections typically have a lifetime measured in minutes and the backup connection is only for the life of the backup operations

This isn't an accurate assumption. It's fairly common for customers to use connection pools with no max connection lifetime. On Serverless, since pods are so volatile, it's likely that connection lifetime is somewhat close to server lifetime.

> Currently the Start function blocks on initial refresh.

I think the map has to be a soft dependency. We can't allow an S3/GCS outage to prevent db nodes from coming online.

andrewbaptist replied:
I agree about the soft dependency, I changed the reload code to only load from the async thread, so it never blocks startup. 

In terms of the long lived connections, as you mentioned, they are somewhat problematic in the sense that they won't pick up any cidr changes. I didn't add this yet, but your suggestion definitely seems doable. It also helps remove any bootstrap issues if there are connections created prior to initialization.

andrewbaptist replied:
This ended up being a bit more complicated than I had originally hoped, but I think the approach I took will work. There are two concerns I wanted to deal with
1) There may be a large number of connections, and when the mapping changes, I didn't want to have thousands of callbacks fire when the mapping changed. 
2) Given a child metric name, it can only be registered once. This means it needs to be centralized in a single map on the server and can't be done "per-connection". 

I took a couple stabs at doing this, but ended up with the current approach. I am definitely open to feedback or other alternatives on how to do this a little better. 

Note that I also updated the CIDR commit to add the SetOnChange method.

Commit:
sql: pass an Inc function to writeBuffer

Previously the writeBuffer required a specific Metric implementation as
part of its paramater. This made it more complicated to change the
metric type that was passed in.

Epic: none

Release note: None

andrewbaptist commented on pkg/sql/pgwire/conn_test.go:
```diff
@@ -742,7 +746,10 @@ func getSessionArgs(
 			// after parsing the client-sent parameters.
 			c := &conn{conn: netConn}
-			c.msgBuilder.init(metric.NewCounter(metric.Metadata{}))
+			aggCounter := aggmetric.NewCounter(MetaBytesIn, "remote")
```
I moved this all back. I had originally thought I needed to look this up later, but that was incorrect.

andrewbaptist commented on pkg/sql/pgwire/write_buffer.go:
```diff
@@ -35,22 +34,16 @@ type writeBuffer struct {
-	// bytecount counts the number of bytes written across all pgwire connections, not just this
-	// buffer. This is passed in so that finishMsg can track all messages we've sent to a network
-	// socket, reducing the onus on the many callers of finishMsg.
-	bytecount *metric.Counter
```
Reverted

andrewbaptist commented:
I ran some system level tests on the metric to validate that it works correctly. 

I validated that it updates end-to-end when the cidr file changes and I also validated that the overhead is low. 
The dynamically update for existing connections and applies the new mappings. The overhead in CPU profiles appears to be ~0.05% for a KV heavy workload  (where it would likely be the highest impact).

jeffswenson reviewed (approved):
LGTM

andrewbaptist commented:
TFTR! 

I removed the LookupURI as you were right and we won't need it. I've modified the backup code to not use it but still need to finish the changefeed. 

bors r=jeffswenson
